### PR TITLE
Add GPU Acceleration apple silicon examples

### LINKED ISSUE #18133: Adding Apple Silicon GPU Acceleration examples
### Description & Motivation

We could add Apple silicon GPU support for the examples of lightning-AI.

### Pitch

Add MPS support for the examples.

### Alternatives

_No response_

### Additional context

Had Created a pull request https://github.com/Lightning-AI/lightning/pull/18127 for this.

cc @borda

### PR DESCRIPTION
**Pull request by NripeshN**
## What does this PR do?
Add GPU Acceleration Apple silicon examples

Fixes #18133

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: examples/app/multi_node/train_pytorch.py [hunk 1]
```diff
@@ -22,7 +22,12 @@ def distributed_train(local_rank: int, main_address: str, main_port: int, num_no
 
     # 2. PREPARE DISTRIBUTED MODEL
     model = torch.nn.Linear(32, 2)
-    device = torch.device(f"cuda:{local_rank}") if torch.cuda.is_available() else torch.device("cpu")
+    if torch.cuda.is_available():
```
Comment (awaelchli) [inline examples/app/multi_node/train_pytorch.py:25]: I would remove MPS from the multi-node scripts here. There is no backend that can support this today, so it's not a realistic use case.

File: examples/app/server/app.py [hunk 2]
```diff
@@ -15,7 +15,12 @@
 class PyTorchServer(PythonServer):
     def setup(self):
         self._model = torchvision.models.resnet18(pretrained=True)
-        self._device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Comment (awaelchli) [inline examples/app/server/app.py:18]: A similar comment here. The cloud compute specified in the app below does not support Apple Silicon hardware.

File: examples/fabric/dcgan/train_torch.py [hunk 3]
```diff
@@ -69,7 +69,12 @@ def main():
     dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=workers)
 
     # Decide which device we want to run on
-    device = torch.device("cuda:0" if (torch.cuda.is_available() and num_gpus > 0) else "cpu")
```
Comment (awaelchli) [inline examples/fabric/dcgan/train_torch.py:72]: These changes look good, thanks!

**Review by awaelchli — APPROVED** [review]
Thanks for sending the update. I have only minor comments

**Review by awaelchli — APPROVED** [review]
Thanks for sending the update. I have only minor comments

File: examples/app/server_with_auto_scaler/app.py [hunk 4]
```diff
@@ -26,7 +26,12 @@ def __init__(self, *args, **kwargs):
         )
 
     def setup(self):
-        self._device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        if torch.cuda.is_available():
```
Comment (Borda) [inline examples/app/server_with_auto_scaler/app.py:29]: As this is used several times, can we wrap it into a function?

Comment (NripeshN) [reply examples/app/server_with_auto_scaler/app.py:29]: Hi, Where should I keep the function?

Comment (Borda) [reply examples/app/server_with_auto_scaler/app.py:29]: seems I was wrong as each example shall be stand-alone...

File: examples/fabric/image_classifier/train_torch.py [hunk 5]
```diff
@@ -56,7 +56,13 @@ def run(hparams):
     torch.manual_seed(hparams.seed)
 
     use_cuda = torch.cuda.is_available()
-    device = torch.device("cuda" if use_cuda else "cpu")
+    use_mps = torch.backends.mps.is_available()
```
Comment (Borda) [inline examples/fabric/image_classifier/train_torch.py:59]: @awaelchli does Fabric has `auto`?

**Review by carmocca — APPROVED** [review]

File: examples/fabric/image_classifier/train_torch.py (hunk 5 above)
Comment (awaelchli) [reply examples/fabric/image_classifier/train_torch.py:59]: Yes, that's what happens in train_fabric.py, since auto is the default.

Comment (Borda) [reply examples/fabric/image_classifier/train_torch.py:59]: So we don't need this, right?

Comment (awaelchli) [reply examples/fabric/image_classifier/train_torch.py:59]: This is the raw pytorch example. It is meant to show how the example looks like in raw pytorch, and next to it we have the train_fabric.py example. 

It is nice to show how much boilerplate code can be removed when using Fabric/Trainer, and this is the motivation behind having it here.
